### ISSUE TITLE
Test connection hangs

### ISSUE THREAD

**Issue by wmorrell** [KostyaSha/yet-another-docker-plugin#154]
Jenkins 2.46.1
jclouds-jenkins 2.14
yet-another-docker-plugin 0.1.0-rc36

I am trying to add in a Docker host; I set the Cloud Name, Docker URL, set credentials, and click the Test Connection button. A spinner and "Testing..." message come up, but the connection never finishes. I also cannot set the Read timeout value; setting it always reverts to an empty value. Nothing shows up in Cloud Statistics, etc.

The cloud section of config.xml is below (IP changed to 192.168.100.100):

    <clouds>
      <com.github.kostyasha.yad.DockerCloud plugin="yet-another-docker-plugin@0.1.0-rc36">
        <name>test Dockerhost</name>
        <provisionedImages/>
        <templates>
          <com.github.kostyasha.yad.DockerSlaveTemplate>
            <id>40665c56-bed1-4139-98c7-49176a8919a9</id>
            <labelString>docker</labelString>
            <launcher class="com.github.kostyasha.yad.launcher.DockerComputerSSHLauncher">
              <sshConnector plugin="ssh-slaves@1.17">
                <port>22</port>
                <credentialsId>66d0997b-fa5b-490c-8246-e846ded7da48</credentialsId>
                <jvmOptions></jvmOptions>
                <javaPath></javaPath>
                <maxNumRetries>0</maxNumRetries>
                <retryWaitTime>0</retryWaitTime>
                <sshHostKeyVerificationStrategy class="hudson.plugins.sshslaves.verifiers.KnownHostsFileKeyVerificationStrategy"/>
              </sshConnector>
            </launcher>
            <remoteFs>/home/jenkins</remoteFs>
            <mode>EXCLUSIVE</mode>
            <retentionStrategy class="com.github.kostyasha.yad.strategy.DockerOnceRetentionStrategy">
              <idleMinutes>10</idleMinutes>
              <idleMinutes defined-in="com.github.kostyasha.yad.strategy.DockerOnceRetentionStrategy">10</idleMinutes>
            </retentionStrategy>
            <numExecutors>1</numExecutors>
            <dockerContainerLifecycle>
              <image>jenkins-agent</image>
              <pullImage>
                <pullStrategy>PULL_NEVER</pullStrategy>
                <credentialsId></credentialsId>
              </pullImage>
              <createContainer>
                <command></command>
                <hostname></hostname>
                <dnsHosts class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </dnsHosts>
                <volumes class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </volumes>
                <volumesFrom class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </volumesFrom>
                <environment class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </environment>
                <bindPorts></bindPorts>
                <bindAllPorts>false</bindAllPorts>
                <memoryLimit>0</memoryLimit>
                <privileged>false</privileged>
                <tty>false</tty>
                <extraHosts class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </extraHosts>
                <devices class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </devices>
                <cpusetCpus></cpusetCpus>
                <cpusetMems></cpusetMems>
                <links class="java.util.Collections$UnmodifiableRandomAccessList" resolves-to="java.util.Collections$UnmodifiableList">
                  <c class="list"/>
                  <list reference="../c"/>
                </links>
              </createContainer>
              <stopContainer>
                <timeout>10</timeout>
              </stopContainer>
              <removeContainer>
                <removeVolumes>true</removeVolumes>
                <force>true</force>
              </removeContainer>
            </dockerContainerLifecycle>
            <nodeProperties/>
            <maxCapacity>10</maxCapacity>
          </com.github.kostyasha.yad.DockerSlaveTemplate>
        </templates>
        <containerCap>50</containerCap>
        <connector>
          <serverUrl>tcp://192.168.100.100:2376</serverUrl>
          <credentialsId>66d0997b-fa5b-490c-8246-e846ded7da48</credentialsId>
          <connectorType>NETTY</connectorType>
          <connectTimeout>5</connectTimeout>
        </connector>
      </com.github.kostyasha.yad.DockerCloud>
    </clouds>

The connection to the remote Docker daemon works fine with the certs from the command line. Output of `docker info` below:

    Containers: 0
     Running: 0
     Paused: 0
     Stopped: 0
    Images: 103
    Server Version: 17.03.1-ce
    Storage Driver: aufs
     Root Dir: /var/lib/docker/aufs
     Backing Filesystem: extfs
     Dirs: 135
     Dirperm1 Supported: true
    Logging Driver: json-file
    Cgroup Driver: cgroupfs
    Plugins: 
     Volume: local
     Network: bridge host macvlan null overlay
    Swarm: inactive
    Runtimes: runc
    Default Runtime: runc
    Init Binary: docker-init
    containerd version: 4ab9917febca54791c5f071a9d1f404867857fcc
    runc version: 54296cf40ad8143b62dbcaa1d90e520a2136ddfe
    init version: 949e6fa
    Kernel Version: 3.16.0-4-amd64
    Operating System: Debian GNU/Linux 8 (jessie)
    OSType: linux
    Architecture: x86_64
    CPUs: 1
    Total Memory: 1.755 GiB
    Name: test-docker
    ID: ZE4T:LL5Z:GOKX:DOB5:C4HD:LXOT:FVUO:YSXC:HFZD:MTNJ:L2ET:FGQW
    Docker Root Dir: /var/lib/docker
    Debug Mode (client): false
    Debug Mode (server): false
    Username: wmorrell
    Registry: https://index.docker.io/v1/
    WARNING: No memory limit support
    WARNING: No swap limit support
    WARNING: No kernel memory limit support
    WARNING: No oom kill disable support
    WARNING: No cpu cfs quota support
    WARNING: No cpu cfs period support
    Experimental: false
    Insecure Registries:
     127.0.0.0/8
    Live Restore Enabled: false

I do not see any errors in Jenkins logs. Any ideas on what else to check?

**Comment by wmorrell** [comment]
Ok, walking away and coming back to this, I realized that the credentials I set for the cloud was the wrong entry; I picked the SSH key for the `authorized_keys` in my container image, instead of the Docker Host Certificates credential. PEBKAC.

For the problem of the Read Timeout field not persisting, I took a peek at the code, and it looks like the `equals` and `hashCode` methods on `com.github.kostyasha.yad.DockerConnector` includes all non-transient fields _except_ for `readTimeout`. I could set the field by manually editing the `config.xml` and reloading config from disk.

Once I get my Jenkins properly configured, I'll see about coming back to this and testing if adding the field to `equals` and `hashCode` fixes the problem with persisting Read Timeout. If it does, I'll make a pull request.
